docs/design/mono/wasm-threads.md

```diff
@@ -0,0 +1,231 @@
+
+## C# Thread
+- could block on synchronization primitives
+- without JS interop. calling JSImport will PNSE.
+
+## C# Threadpool Thread
+- could block on synchronization primitives
+- without JS interop. calling JSImport will PNSE.
+
+## JSImport and marshaled JS functions
+- both sync and async could be called on all threads
+- sync: when called from C# it will use `SynchronizationContext.Send` and block caller.
+- async: when called from C# it will use `SynchronizationContext.Post` and marshal promise immediately.
+- when this is worker -> worker, `SynchronizationContext` should invoke it inline
+
+## JSExport & C# delegates
+- sync: will throw PNSE if called from UI JavaScript
+- sync: will just work when called from JSWebWorker JavaScript
+- async JSExport: will work on all threads. Will marshal promise and return immediately.
+- async Delegate: are there any async callback possible yet ? The code gen doesn't support it yet in Net8.
+- `getAssemblyExports` need to bind JS on UI thread, but register on deputy thread
+- hide `SynchronizationContext.Send` and `SynchronizationContext.Post` inside of the generated code.
+    - fast on worker -> worker
+
+## Promise
+- passing promise should work everywhere.
+- from UI javaScript it would be passed as Task to deputy worker
+- open question: passing JS promise to deputy should be fine. But does the `resolve()` need to block UI thread ?
+
+## Task, Task<T>
+- passing Task should work everywhere.
+- when marshaled to JS they bind to specific Promise and have affinity
+    - the `SetResult` need to be marshaled on thread of the Promise.
+    - The proxy of the Promise knows which `SynchronizationContext` to dispatch to.
+    - on UI thread it's the UI thread's SynchronizationContext, not deputy's
+    - TODO: could same task be marshaled to multiple JS workers ?
+
+## JSObject proxy
+- has thread affinity, marked by private ThreadId.
+    - in deputy worker, it will be always UI thread Id
+    - the JSHandle always belongs to UI thread
+- `Dispose` need to be called on the right thread.
+    - how to do that during GC/finalizer ?
+    - should we run finalizer per worker ?
+- is it ok for `SynchronizationContext` to be public API
+    - because it could contain UI thread SynchronizationContext, which user code should not be disposed on.
+
+## JSHost.GlobalThis, JSHost.DotnetInstance, JSHost.ImportAsync
+- calls will be dispatched from deputy thread to UI JavaScript
+- on JSWebWorker call will stay on the same thread.
+
+## SynchronizationContext
+- we will need public C# API for it, `JSHost.xxxSynchronizationContext`
+    - we could avoid it by generating late bound ICall. Very ugly.
+- hide `SynchronizationContext.Send` and `SynchronizationContext.Post` inside of the generated code.
+    - needs to be also inside generated nested marshalers
+    - is solution for deputy's SynchronizationContext same as for JSWebWorker's SynchronizationContext, from the code-gen perspective ?
```

Comment:
I want the answer to be "yes"

Reply:
I hope so too. Over night I realized that shape of this API could be 2 methods on the `JSHost` rather than exposing the `SynchronizationContext`, because there could more complex logic on which of the should be used.